[iPad] Feedly App doesn't respond to left/right swipe gestures with trackpad

whsieh opened the pull request:
#### 70523aab320c9c768da1c44d5da7f6d983f5057a
<pre>
[iPad] Feedly App doesn&apos;t respond to left/right swipe gestures with trackpad
<a href="https://bugs.webkit.org/show_bug.cgi?id=282539">https://bugs.webkit.org/show_bug.cgi?id=282539</a>
<a href="https://rdar.apple.com/138044440">rdar://138044440</a>

Reviewed by Aditya Keerthi.

When using a trackpad to try and swipe horizontally between articles in the Feedly app on iPad,
nothing happens. In contrast, a horizontal pan gesture via direct touch works as expected. This is
because Feedly mostly wraps a web view which loads nearly all of their app&apos;s content; when viewing
an article, touch event listeners are used to detect when the user swipes horizontally, and then run
an animated transition to load the next or previous article. Because touch events are only
dispatched in response to actual digitizer events (as opposed to indirect touches via trackpad),
swiping horizontally with a trackpad results in nothing.

Ideally, Feedly would listen for `wheel` events and use similar logic to transition between articles
in this case. However, in lieu of the app adopting wheel events or a pan/swipe gesture recognizer,
we can add a quirk to make this work by automatically emitting a stream of touch events in response
to scrolling triggered by a 2-finger swipe on a trackpad. To achieve this, we add a new helper class
(`PointerTouchCompatibilitySimulator`) that&apos;s responsible for:

1.  Detecting mostly-horizontal trackpad scrolling deltas, and then
2.  Using `_WKTouchEventGenerator` (currently used to simulate touches for WebDriver on iOS) to
    synthesize touch events in the direction of the swipe, starting at the pointer location.

See below for more details.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _initializeWithConfiguration:]):

Create a `PointerTouchCompatibilitySimulator` if needed.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView scrollView:handleScrollUpdate:completion:]):

Deliver scroll updates to the `PointerTouchCompatibilitySimulator`.

* Source/WebKit/UIProcess/API/Cocoa/_WKTouchEventGenerator.h:
(-[_WKTouchEventGenerator touchDown:window:]):
(-[_WKTouchEventGenerator touchDown:window:completionBlock:]):
(-[_WKTouchEventGenerator liftUp:window:]):
(-[_WKTouchEventGenerator liftUp:window:completionBlock:]):
(-[_WKTouchEventGenerator moveToPoint:duration:window:]):
(-[_WKTouchEventGenerator moveToPoint:duration:window:completionBlock:]):

Refactor the touch event generator, exposing versions of several methods for synthesizing touch
events without requiring a completion handler (and consequently, marker HID events). This is because
marker events require the `UIApplication` subclass to explicitly use SPI to deliver HID events to
the touch event generator.

* Source/WebKit/UIProcess/API/Cocoa/_WKTouchEventGeneratorInternal.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKTouchEventGenerator.h.

Add an internal header for `_WKTouchEventGenerator` where the new helper methods above are declared.

* Source/WebKit/UIProcess/ios/PointerTouchCompatibilitySimulator.h: Copied from Source/WebKit/UIProcess/_WKTouchEventGenerator.h.
(WebKit::PointerTouchCompatibilitySimulator::view const):
(WebKit::PointerTouchCompatibilitySimulator::window const):
* Source/WebKit/UIProcess/ios/PointerTouchCompatibilitySimulator.mm: Added.
(WebKit::PointerTouchCompatibilitySimulator::requiresPointerTouchCompatibility):
(WebKit::PointerTouchCompatibilitySimulator::PointerTouchCompatibilitySimulator):
(WebKit::hasPredominantHorizontalAxis):
(WebKit::PointerTouchCompatibilitySimulator::handleScrollUpdate):

Add a helper class to receive scroll updates and detect when scroll updates are mostly horizontal
(in the future, this could be extended to vertical directions, and even pinch or rotation gestures).
When a predominantly horizontal scroll deltas are detected, we then begin synthesizing touch events,
starting at the `-locationInView:` and moving away from this start location based on the values of
`-translationInView:`. Note that we also require a watchdog timer to ensure that we don&apos;t leave the
app in a state where pointer gestures are permanently stuck, by always ending the synthetic touch
event stream after ~50 ms with no further scrolling. This isn&apos;t necessary most of the time, because
we receive `BEScrollViewScrollUpdatePhase(Ended|Cancelled)` and immediately end the event stream,
but in practice it&apos;s possible that we never get this final update with ended/cancelled phase when
performing a swipe.

(WebKit::PointerTouchCompatibilitySimulator::resetState):
(WebKit::PointerTouchCompatibilitySimulator::locationInScreen const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/286150@main">https://commits.webkit.org/286150@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/225f20cc24fa7b800f47653b1960f9bea13fadfe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74967 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54397 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27784 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79397 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26206 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63533 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2182 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58853 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17127 "Found 2 new test failures: webgl/2.0.y/conformance/context/context-release-with-workers.html webgl/2.0.y/conformance/rendering/canvas-alpha-bug.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78034 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49008 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64405 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39244 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46312 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24538 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67453 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22248 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80883 "") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2285 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1381 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/80883 "") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2434 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64424 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/80883 "") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16507 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10352 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8514 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2250 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5038 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2278 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3199 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2285 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -455,6 +456,11 @@ - (void)_initializeWithConfiguration:(WKWebViewConfiguration *)configuration
 #endif
+
+#if PLATFORM(IOS_FAMILY)
+    if (WTF::IOSApplication::isFeedly())
```
Even though it's just one app for now, I wonder if we should abstract this into a method? `needsPointerTouchCompatibility`?

Just seems slightly strange to have the application check in directly in the initializer 😅

whsieh replied:
Good point — I'll move this into a `static bool requiresPointerTouchCompatibility();` method on the `PointerTouchCompatibilitySimulator` class instead, and consolidate logic for touch simulation there.

pxlcoder commented on Source/WebKit/UIProcess/ios/PointerTouchCompatibilitySimulator.mm:
```diff
@@ -0,0 +1,112 @@
+    if (scrollView._wk_canScrollHorizontallyWithoutBouncing)
+        return;
+
+    auto translation = [update translationInView:view.get()];
```
Looks like `translationInView` is not available on visionOS?

whsieh replied:
Ah, in the `!USE(BROWSERENGINEKIT)` codepath the scroll update is just a `UIScrollEvent` (SPI), which uses `-_adjustedAcceleratedDeltaInView:`. Will change to that.

pxlcoder commented on Source/WebKit/UIProcess/ios/PointerTouchCompatibilitySimulator.mm:
```diff
@@ -0,0 +1,112 @@
+
+#if PLATFORM(IOS_FAMILY)
+
+#import "UIKitSPI.h"
```
Is SPI usage avoidable? I could not obviously spot its use.

whsieh replied:
Oh, you're right — it's not used in this PR. It'll be needed once I fix the visionOS build, though :/

whsieh committed:
[iPad] Feedly App doesn't respond to left/right swipe gestures with trackpad
https://bugs.webkit.org/show_bug.cgi?id=282539
rdar://138044440

Reviewed by Aditya Keerthi.

When using a trackpad to try and swipe horizontally between articles in the Feedly app on iPad,
nothing happens. In contrast, a horizontal pan gesture via direct touch works as expected. This is
because Feedly mostly wraps a web view which loads nearly all of their app's content; when viewing
an article, touch event listeners are used to detect when the user swipes horizontally, and then run
an animated transition to load the next or previous article. Because touch events are only
dispatched in response to actual digitizer events (as opposed to indirect touches via trackpad),
swiping horizontally with a trackpad results in nothing.

Ideally, Feedly would listen for `wheel` events and use similar logic to transition between articles
in this case. However, in lieu of the app adopting wheel events or a pan/swipe gesture recognizer,
we can add a quirk to make this work by automatically emitting a stream of touch events in response
to scrolling triggered by a 2-finger swipe on a trackpad. To achieve this, we add a new helper class
(`PointerTouchCompatibilitySimulator`) that's responsible for:

1.  Detecting mostly-horizontal trackpad scrolling deltas, and then
2.  Using `_WKTouchEventGenerator` (currently used to simulate touches for WebDriver on iOS) to
    synthesize touch events in the direction of the swipe, starting at the pointer location.

See below for more details.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _initializeWithConfiguration:]):

Create a `PointerTouchCompatibilitySimulator` if needed.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView scrollView:handleScrollUpdate:completion:]):

Deliver scroll updates to the `PointerTouchCompatibilitySimulator`.

* Source/WebKit/UIProcess/API/Cocoa/_WKTouchEventGenerator.h:
(-[_WKTouchEventGenerator touchDown:window:]):
(-[_WKTouchEventGenerator touchDown:window:completionBlock:]):
(-[_WKTouchEventGenerator liftUp:window:]):
(-[_WKTouchEventGenerator liftUp:window:completionBlock:]):
(-[_WKTouchEventGenerator moveToPoint:duration:window:]):
(-[_WKTouchEventGenerator moveToPoint:duration:window:completionBlock:]):

Refactor the touch event generator, exposing versions of several methods for synthesizing touch
events without requiring a completion handler (and consequently, marker HID events). This is because
marker events require the `UIApplication` subclass to explicitly use SPI to deliver HID events to
the touch event generator.

* Source/WebKit/UIProcess/API/Cocoa/_WKTouchEventGeneratorInternal.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKTouchEventGenerator.h.

Add an internal header for `_WKTouchEventGenerator` where the new helper methods above are declared.

* Source/WebKit/UIProcess/ios/PointerTouchCompatibilitySimulator.h: Copied from Source/WebKit/UIProcess/_WKTouchEventGenerator.h.
(WebKit::PointerTouchCompatibilitySimulator::view const):
(WebKit::PointerTouchCompatibilitySimulator::window const):
* Source/WebKit/UIProcess/ios/PointerTouchCompatibilitySimulator.mm: Added.
(WebKit::PointerTouchCompatibilitySimulator::requiresPointerTouchCompatibility):
(WebKit::PointerTouchCompatibilitySimulator::PointerTouchCompatibilitySimulator):
(WebKit::hasPredominantHorizontalAxis):
(WebKit::PointerTouchCompatibilitySimulator::handleScrollUpdate):

Add a helper class to receive scroll updates and detect when scroll updates are mostly horizontal
(in the future, this could be extended to vertical directions, and even pinch or rotation gestures).
When a predominantly horizontal scroll deltas are detected, we then begin synthesizing touch events,
starting at the `-locationInView:` and moving away from this start location based on the values of
`-translationInView:`. Note that we also require a watchdog timer to ensure that we don't leave the
app in a state where pointer gestures are permanently stuck, by always ending the synthetic touch
event stream after ~50 ms with no further scrolling. This isn't necessary most of the time, because
we receive `BEScrollViewScrollUpdatePhase(Ended|Cancelled)` and immediately end the event stream,
but in practice it's possible that we never get this final update with ended/cancelled phase when
performing a swipe.

(WebKit::PointerTouchCompatibilitySimulator::resetState):
(WebKit::PointerTouchCompatibilitySimulator::locationInScreen const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/286150@main
